### PR TITLE
pinns: Allow pinning mount namespaces

### PR DESCRIPTION
**Pull request by lack**
Mount namespaces are a bit odd in that:
- Like user namespaces, you must fork and bindmount the mount namespace
  from the parent process (otherwise the bindmount is in the child
  mount namespace where no one sees it)
- Additionally, you cannot bindmount a mount namespace to a file if that
  file is in a shareable mount namespace, otherwise infinite loops are
  possible. So before attempting to bind the mount namespace we
  bind-mount the target directory to itself (if it's not already a
  mountpoint) and mark it as unbindable.

Usage:
    pinns -d ... -f ... [-m|--mnt]

Signed-off-by: Jim Ramsay <jramsay@redhat.com>

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This adds the ability for the pinns utility to pin mount namespaces.  This will be useful in the implementation of My [hide-container-mountpoints](https://github.com/lack/openshift-enhancements/blob/hide_container_mountpoints/enhancements/hide_container_mountpoints.md) Openshift enhancement proposal.

#### Which issue(s) this PR fixes:

None

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Review by saschagrunert — COMMENTED** [review]
/ok-to-test

**Review by saschagrunert — APPROVED** [review]
LGTM, I think we're at a point where we would need some tests for those code lines.

**Comment by lack** [comment]
@saschagrunert I didn't see any tests about pinns specifically.  Can you point me to where I might add some?

**Comment by saschagrunert** [comment]
> @saschagrunert I didn't see any tests about pinns specifically. Can you point me to where I might add some?

There are none right now, we may follow-up with that in another PR.

File: pinns/src/pinns.c [hunk 1]
```diff
@@ -268,20 +294,62 @@ static bool is_host_ns(const char* const optarg) {
...
+  if (create_bind_root(bind_root, PATH_MAX - 1, pin_path, ns_name) < 0) {
+    return -1;
+  }
+
+  // TODO: Check if bind_root is already a mountpoint
+  // For now, just blindly try to bindmount itself to itself.
+  // If this succeeds, we know it's a mountpoint.  If it fails it probably
+  // failed because it's already a mountpoint, and if it's not, the call to set
+  // MS_UNBINDABLE will fail next.
+  if (mount(bind_root, bind_root, NULL, MS_BIND, NULL) < 0) {
```
Comment (haircommander) [inline pinns/src/pinns.c:325]: it seems we're not reacting to mount()<0. Why have the if statement?

File: pinns/src/pinns.c [hunk 2]
```diff
@@ -268,20 +294,62 @@ static bool is_host_ns(const char* const optarg) {
...
+/* Mount namespaces can only be bound into unsharable mount namespaces (to
+ * avoid infinite loops), so force the pin_path to be a bind-mount to
+ * itself and then marked as unsharable. */
+static int setup_unbindable_bindpath(const char *pin_path, const char *ns_name) {
+  char bind_root[PATH_MAX];
+
+  // first, verify the /$PATH/$NSns directory exists
+  if (create_bind_root(bind_root, PATH_MAX - 1, pin_path, ns_name) < 0) {
+    return -1;
+  }
```
Comment (haircommander) [inline pinns/src/pinns.c:318]: can these functions go beneath their last caller?

Comment (haircommander) [reply pinns/src/pinns.c:318]: ah I see setup_unbindable_bindpath is, but create_bind_root could be moved to the bottom of the file

File: pinns/src/pinns.c [hunk 3]
```diff
@@ -196,25 +208,29 @@ int main(int argc, char **argv) {
...
 
-    if (gid_mapping && write_mapping_file(pid, gid_mapping, true) < 0)
-      pexit("Cannot write gid mappings");
+      /* Write user mappings */
+      if (gid_mapping && write_mapping_file(pid, gid_mapping, true) < 0)
+        pexit("Cannot write gid mappings");
 
-    if (uid_mapping && write_mapping_file(pid, uid_mapping, false) < 0)
-      pexit("Cannot write gid mappings");
+      if (uid_mapping && write_mapping_file(pid, uid_mapping, false) < 0)
```
Comment (haircommander) [inline pinns/src/pinns.c:225]: nit: there seem to be some whitespace changes I wouldnt' expect (not sure if they're intentional)

**Comment by haircommander** [comment]
@giuseppe PTAL

File: pinns/src/pinns.c (hunk 3 above)
Comment (lack) [reply pinns/src/pinns.c:225]: Intentional: These were indented into the `if (bind_user) {` on [line 215 above](https://github.com/cri-o/cri-o/pull/4508/files/60986224f0d5dd4879aef283a5713e308e0533bd#diff-bfef5beff459416de7e125890a18464df27f7ada77aa3d6f89eb033fac790801R215).

File: pinns/src/pinns.c (hunk 2 above)
Comment (lack) [reply pinns/src/pinns.c:318]: Will do.

File: pinns/src/pinns.c (hunk 1 above)
Comment (lack) [reply pinns/src/pinns.c:325]: 2 reasons: A lot of coverage tools get cranky if you don't check the return value in some way, and I wanted to make it explicit that we're explicitly ignoring errors.

Happy to do it some other way.

Comment (lack) [reply pinns/src/pinns.c]: Trying to just ignore it to see if any tooling here complains...

**Comment by haircommander** [comment]
note: ci failures aren't related https://github.com/cri-o/cri-o/issues/4509

File: pinns/src/pinns.c [hunk 4]
```diff
@@ -268,20 +295,48 @@ static bool is_host_ns(const char* const optarg) {
...
+  // first, verify the /$PATH/$NSns directory exists
+  if (create_bind_root(bind_root, PATH_MAX - 1, pin_path, ns_name) < 0) {
+    return -1;
+  }
+
+  // TODO: Check if bind_root is already a mountpoint
+  // For now, just blindly try to bindmount itself to itself, ignoring
+  // failures.  If this succeeds, we know it's a mountpoint.  If it fails it
+  // probably failed because it's already a mountpoint, and if it's not, the
+  // call to set MS_UNBINDABLE will fail next.
```
Comment (giuseppe) [inline pinns/src/pinns.c:313]: could we just attempt an `umount` before the `mount` (or even better run in a loop until it succeeds)?  Just to make sure we don't pile up mounts if it runs multiple times.

**Comment by giuseppe** [comment]
one note for `Hide Container Mountpoints`.  Moving CRI-O to a separate mount namespace will affect also Podman/Skopeo/Buildah, and they should be running from the same mount namespace, otherwise it could cause issues in the underlying storage.

File: pinns/src/pinns.c (hunk 4 above)
Comment (lack) [reply pinns/src/pinns.c:313]: I think doing an `umount` and then mount again would be problematic if `bind_root` is already a mountpoint we don't manage, or if we're reusing the same bind_root to add an additional mount namespace pin to it.  The behavior right now is that if `bind_root` is already mounted (including if it's already a bind-mount to itself like we intend to do the 1st time around), this mount call fails and there's nothing that piles up.

I was looking around for a nice way to check if it's already mounted, but it looked like the only way is to parse and walk through /proc/self/mount[s|info], or to use something like libmount that does this already, and either way it seemed heavy-weight when a try-and-fail-if-it's already-mounted does the job okay.

**Comment by lack** [comment]
FYI: I've done some manual testing of this change, including double-checking that 'nsenter --mount=...' can indeed use the mount namespace file pinned in this way.

File: pinns/src/pinns.c (hunk 4 above)
Comment (giuseppe) [reply pinns/src/pinns.c:313]: > I was looking around for a nice way to check if it's already mounted, but it looked like the only way is to parse and walk through /proc/self/mount[s|info], or to use something like libmount that does this already, and either way it seemed heavy-weight when a try-and-fail-if-it's already-mounted does the job okay.

agreed.  It is not worth to add that complexity

**Comment by haircommander** [comment]
LGTM

how do you feel about this @giuseppe

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by haircommander** [comment]
/lgtm

thanks @lack

**Comment by lack** [comment]
I rebase this to the tip of the current master branch, to try to unstick that CI job...

**Comment by lack** [comment]
@haircommander : This will need a re-lgtm thanks to the rebase.

**Pull request merged**